### PR TITLE
Perform a null check for a pattern-based fixed statement over nullable value type.

### LINKED ISSUE #66167: Pattern-based fixed statement behaves inconsistently for nullable value types
```
unsafe class C
{
    public static void Main()
    {
        Test((int?)null);
        Test2((int?)null);
    }

    public static void Test<T>(T arg)
    {
        fixed (int* p = arg)
        {
            System.Console.Write(p == null? 0: p[1]);
        }
    }
    public static void Test2(int? arg)
    {
        fixed (int* p = arg)
        {
            System.Console.Write(p == null? 0: p[1]);
        }
    }
}

static class FixAllExt
{
    public static ref int GetPinnableReference<T>(this T dummy)
    {
        return ref (new int[]{1,2,3})[0];
    }
}
```

Observed:
```
02
```

Expected:
Consistent behavior between generic and not generic code. The output should be either ```00```, or ```22```.

It looks like https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.3/pattern-based-fixed.md#translation implies that there must be a null check for the target, but it is not clear whether the intent was to actually perform the null check for nullable value types.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #66167

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs [hunk 1]
```diff
@@ -329,15 +329,25 @@ public override BoundNode VisitFixedLocalCollectionInitializer(BoundFixedLocalCo
             BoundExpression callReceiver;
             int currentConditionalAccessID = 0;
 
-            bool needNullCheck = !initializerType.IsValueType;
-
+            bool needNullCheck = !initializerType.IsValueType || initializerType.IsNullableType();
+            BoundAssignmentOperator? nullableTempAssignment = null;
+            BoundLocal? nullableBoundTemp = null;
+
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs:335]:
```suggestion

```

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs [hunk 2]
```diff
@@ -329,15 +329,24 @@ public override BoundNode VisitFixedLocalCollectionInitializer(BoundFixedLocalCo
             BoundExpression callReceiver;
             int currentConditionalAccessID = 0;
 
-            bool needNullCheck = !initializerType.IsValueType;
-
+            bool needNullCheck = !initializerType.IsValueType || initializerType.IsNullableType();
+            BoundAssignmentOperator? nullableTempAssignment = null;
+            BoundLocal? nullableBoundTemp = null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs:334]:
```suggestion
            BoundLocal? nullableBoundTemp = null;

```

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

**Review by jjonescz — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Review by 333fred — APPROVED** [review]

**Pull request merged**